torchbraid/torchbraid_app.pyx

```diff
@@ -440,4 +504,120 @@ class BraidApp:
     cdef braid_Core core = (<PyBraid_Core> self.py_core).getCore()
     return (core.grids[0].ilower, core.grids[0].iupper)
 
+
+  def interp_network_state(self, coarse_app, cf):
+    
+    ''' 
+    Interp the network state present in coarse_app into this app.  It is
+    assumed that this app is a regular refinement (relative to coarsening
+    factor cf) of coarse_app.  
+    '''
+    
+
+    # Get braid_Core
+    cdef braid_Core fine_core = (<PyBraid_Core> self.py_core).getCore()
+    cdef braid_Core coarse_core = (<PyBraid_Core> coarse_app.py_core).getCore()
+    
+    # Get fine and coarse grid information.  Note, braid traditionally uses
+    # values like ncpoints, clower, and cupper for these calcuations, but we
+    # don't do that here, because we want to treat the sequential Braid and
+    # MGRIT Braid modes the same.
+    cdef fine_ilower = fine_core.grids[0].ilower    
+    cdef fine_iupper = fine_core.grids[0].iupper    
+    cdef coarse_ilower = coarse_core.grids[0].ilower    
+    cdef coarse_iupper = coarse_core.grids[0].iupper    
+    #print(fine_ilower, fine_iupper, coarse_ilower, coarse_iupper)
+
+    cdef braid_BaseVector fine_bv, coarse_bv
+
+    # The number of C points in fine app (self) must match the number of points in the coarse_app
+    fine_indices = np.arange(fine_ilower, fine_iupper+1)
+    fine_indices = fine_indices[ (fine_indices % cf) == 0 ] # picks off the Cpts 
+    coarse_indices = np.arange(coarse_ilower, coarse_iupper+1)
+    #print(fine_indices, coarse_indices)
+    assert( fine_indices.shape[0] == coarse_indices.shape[0])
+
+    ##
+    # Note, this is only designed to work in serial.  When moving to parallel,
+    # we'll have to send messages and use GetProc
+    ## 
+
+    for fine_pt, coarse_pt in zip(fine_indices, coarse_indices):
+      
+      # Get the fine vector
+      _braid_UGetVectorRef(fine_core, 0, fine_pt, &fine_bv)
+      assert(<unsigned int>(fine_bv)!=0)
+
+      # Get the coarse vector
+      _braid_UGetVectorRef(coarse_core, 0, coarse_pt, &coarse_bv)
+      assert(<unsigned int>(coarse_bv)!=0)
+ 
+      # Copy coarse to fine at C-pt
+      # --> ERIC please check how I do this deep copy from coarse_bv --> fine_bv 
+      py_fine_bv = <object> fine_bv.userVector
+      py_coarse_bv = <object> coarse_bv.userVector
+      py_coarse_bv_clone = BraidVector(torch.clone(py_coarse_bv.tensor_data_), 0)
```

Comment:
We might do `py_coarse_bv.tensor_data_.detach().clone()` here. The reason is that clone has semantics with regard to automatic differentiation. The call to detach separates the tensor from the AD graph, and then. allows a clean clone.

Reply:
Done